### DISCUSSION TITLE
hope future versions can add “form functionality”

### DISCUSSION THREAD

**Discussion by xiaozhiagi** [zauberzeug/nicegui#1242] [Ideas / Feature Requests]
### Description

 I am currently implementing a function like func(param1, param2, param3) for interactive input on a webpage. I hope to be able to run this function on the backend and return the results by clicking a button. Although on_change can trigger real-time updates on the page, I have found it difficult to retrieve the real-time values of items like number, input, etc. on the backend. It’s inconvenient to retrieve multiple interactive values, so I hope future versions can add “form functionality” and consider how to retrieve or manually submit forms in real-time during multi-parameter interactions. Of course, it’s also possible that I’m not very familiar with the usage of nicegui at the moment, so I appreciate any guidance. Thank you!

**Comment by falkoschindler** [comment]
Can you clarify what you're trying to do?
To me it looks like you just need some `ui.input`, `ui.number` etc. and a `ui.button` "Submit". Once the button is clicked you call whatever function you wish and pass the current values of the input elements.

**Reply by xiaozhiagi** [reply]
Your understanding is correct. I have already solved the issue. I used a global variable args_dict to receive the real-time input values from ui.number, and other inputs. The following code is my demo:

```python
# coding: utf-8
from nicegui import app, ui

args_dict = {}

def func_(param1, param2, param3):
    a = param3 * param2
    b = param1
    return a, b

param = {
    "param1": {'label': '参数1', 'name': 'param1', 'htype': 'int', 'value': 5.0},
    'param2': {'label': '参数2', 'name': 'param2', 'htype': 'float', 'value': 2.0},
    'param3': {'label': '参数3', 'name': 'param3', 'htype': 'int', 'value': 15.0},
}

def generate_input_widget2(name, label, htype, value=None, **kwargs):
    def ff(x):
        args_dict[name] = x.value

    if htype == "float":
        return ui.number(label=label, value=value, format='%f', on_change=lambda e: ff(e)).style(
            "margin-left:10%;width:80%;font-size:15px").value
    elif htype == "int":
        return ui.number(label=label, value=int(value), format='%d', on_change=lambda e: ff(e)).style(
            "margin-left:10%;width:80%;font-size:15px").value

for arg in param.keys():
    args_dict[arg] = generate_input_widget2(**param[arg])

def run_():
    res = func_(**args_dict)
    ui.markdown("结果result：" + str(res)).style("margin-left:10%;width:80%;font-size:15px")
    print(res)

ui.button("运行run", on_click=lambda: run_()).style("margin-left:10%;font-size:15px")

ui.run()
```

**Comment by rodja** [comment]
How would a API to simplify this look like? We have a ongoin discussion about form validation in https://github.com/zauberzeug/nicegui/discussions/801. Is this related? Is there a way to solve both feature requests with a single new API?

**Comment by rodja** [comment]
@natankeddem has posted two new approaches in https://github.com/zauberzeug/nicegui/discussions/1791.
